src/kOS.Safe/Compilation/KS/Compiler.cs

```diff
@@ -2024,6 +2030,49 @@ private bool VarIdentifierEndsWithSuffix(ParseNode node)
                  prevChild.Token.Type == TokenType.suffixterm);
         }
 
+        ///<summary>
+        /// Get the rightmost suffix in a var_identifier.
+        /// i.e. return "BBB" if the var_identifier is:<br/>
+        ///    AAA:BBB, or<br/>
+        ///    AAA[0]:BBB,<br/>
+        /// </summary>
+        private string GetVarIdentifierEndSuffix(ParseNode node)
+        {
+            // If it's a var_identifier being worked on, drop down one level first
+            // to get into the actual meat of the syntax tree it represents:
+            if (node.Token.Type == TokenType.varidentifier)
+                return GetVarIdentifierEndSuffix(node.Nodes.First());
+
+            // Descend the rightmost children until encountering the deepest node that is
+            // still a suffix_trailer, array_trailer, or function_trailer.  If that node
+            // was a suffix_trailer, return true, else it's false.
+            ParseNode prevChild = node;
+            ParseNode thisChild = node.Nodes.Last();
+
+            bool descendedThroughAColon = false; // eeeewwww, sounds disgusting.
+
+            while (thisChild.Token.Type == TokenType.suffix_trailer ||
+                   thisChild.Token.Type == TokenType.suffix || 
+                   thisChild.Token.Type == TokenType.suffixterm || 
+                   thisChild.Token.Type == TokenType.suffixterm_trailer || 
+                   thisChild.Token.Type == TokenType.array_trailer ||
+                   thisChild.Token.Type == TokenType.function_trailer)
+            {
+                if (thisChild.Token.Type == TokenType.suffix_trailer)
+                    descendedThroughAColon = true;
+                prevChild = thisChild;
+                thisChild = thisChild.Nodes.Last();
+            }
+            if (descendedThroughAColon && 
+                (prevChild.Token.Type == TokenType.suffix_trailer ||
+                 prevChild.Token.Type == TokenType.suffixterm))
+            {
+                return GetIdentifierText(prevChild);
+            }
+
+            throw new KOSYouShouldNeverSeeThisException("VarIdentifier didn't end in a suffix");
+        }
```

Comment:
It occurs to me that ``GetVarIdentifierEndSuffix()`` and ``GetVarIdentifierEndsWithSuffix()`` are almost identical cut and paste code.  Can we avoid repeated code here by having them share a commonality?  Both perform the same walk.  They only differ in what they do when they get to the bottom.  One returns the string and throws exception if there's no such string to return.  The other sees if there would be such a string and returns boolean whether or not there was.

These are almost exactly the same thing.

Could we have ``GetVarIdentifierEndSuffix()`` just return null (or ``""``) when there's no string (as opposed to throwing a never-see-this exception), and then completely gut the insides of ``GetVarIdentifierEndsWithSuffix()`` to have it just call ``GetVarIdentifierEndSuffix()`` and return true or false depending on whether a string was returned?